src/lib/base.jl

```diff
@@ -122,13 +122,23 @@ end
     function pairs_namedtuple(Δ::Dict)
         t0 = map(zero, t)
         for (idx, v) in Δ
-            t0 = NamedTuple{N}(Base.setindex((t0...,), v, idx))
+            ii = idx isa Integer ? idx : findfirst(==(idx), keys(t))
+            t0 = NamedTuple{N}(Base.setindex((t0...,), v, ii))
         end
         return (t0,)
     end
     return pairs(t), pairs_namedtuple
 end
 
+# For merge between NamedTuple and Dict, we will just convert the Dict to a NamedTuple.
+# and then call `pullback`, which should overall be pretty efficient code generated,
+# and it avoids trying to AD the problematic generic `merge(::NamedTuple, ::iter)` method which uses `push!`.
+if VERSION >= v"1.6"
+  @adjoint merge(nt::NamedTuple, dict::Dict) = pullback(merge, nt, NamedTuple(dict))
+else
+  @adjoint merge(nt::NamedTuple, dict::Dict) = pullback(merge, nt, (;dict...))
```

Comment:
Just keep this?

Reply:
What do you mean?
You mean use if for both 1.6 and pre-1.6?

Reply:
Yeah, exactly.

Reply:
ok